edx_prefectutils/amplitude.py

```diff
@@ -0,0 +1,347 @@
+
+        # Finally, batch the events such that each batch is small enough to fit inside a single Amplitude API call.
+        yield batch_generator(events_for_next_worker, batch_size)
+
+
+def _get_old_high_watermark(sf_credentials: SFCredentials, sf_role: str):
+    """
+    Get the amplitude event ID of the event immediately following the very last event sent to Amplitude's API.
+
+    Raises:
+        snowflake.connector.ProgrammingError: if the loader log table is inaccessible for some reason.
+    """
+    sf_connection = create_snowflake_connection(sf_credentials, sf_role)
+    query = """
+        select max(amplitude_event_id)
+          from prod.amplitude_events.amplitude_events_loader_log
+        """
+    cursor.execute(query)
+    old_high_watermark = cursor.fetchone()[0]
+    if not old_high_watermark:
+        old_high_watermark = 0
+
+
+def _get_new_high_watermark(sf_credentials: SFCredentials, sf_role: str):
+    """
+    TODO: make this more robust (e.g. what if the table does not exist?  what if the table is empty?)
+    """
+    sf_connection = create_snowflake_connection(sf_credentials, sf_role)
+    cursor = sf_connection.cursor()
+    # Get the current maximum amplitude_event_id.
+    query = """
+        select max(amplitude_event_id)
+          from prod.amplitude_events.amplitude_events
+        """
+    cursor.execute(query)
+    return cursor.fetchone()[0]
+
+
+def get_processing_segments():
+    # Get the old (i.e. current) high watermark.
+    old_high_watermark = _get_old_high_watermark()
+
+    # Get the new high watermark, i.e. the current maximum amplitude_event_id.
+    new_high_watermark = _get_new_high_watermark(sf_credentials, sf_role)
+
+    # Get a list of processing segments which cover all events between the old and
+    # new high watermark. The result is a list of start/end tuples describing
+    # amplitude_event_id ranges, where each range represents a single processing
+    # segment with the configured batch size.  The start ID is inclusive, whereas
+    # the end ID is exclusive.
+    #
+    # e.g. if old=1000000, new=3600000, bucketsize=1000000, then:
+    # processing_segments = [
+    #     (1000000, 2000000),
+    #     (2000000, 3000000),
+    #     (3000000, 3600001),
+    # ]
+    #
+    # In the above example, the last batch is smaller than 1m events because there
+    # are no more events to include.
+    return [
+        (processing_segment_start, min(processing_segment_start + processing_bucket_size, new_high_watermark + 1))
+        for processing_segment_start in range(old_high_watermark, new_high_watermark, processing_bucket_size)
+    ]
+
+
+def load_events_into_amplitude(amplitude_api_key: str, processing_segments: list)
+    # Iterate over processing segments in chronological order, handling each one
+    # one at a time, and completely freeing the memory of the previous one before
+    # fetching the next.
+    for processing_segment in processing_segments:
```

Comment:
Each iteration of this loop could become a task, as long as the tasks can be made to run sequentially.  This eliminates a big reason to use tasks which in my mind is automatic concurrency and dependency resolution.  However, it may still be useful to make this into tasks for better visibility into the current progress, and ability to do things like mark a task run as "skipped" or "failed" if that's ever a possible scenario, or "retried" if we want to see retries manifest as task retries instead of backoff retries hidden in the logs.

Reply:
But if we are going to go on a mark tasks as "skipped" or "failed", then we are also not running them sequentially.